src/main/scala/PropertyChecker.scala

```diff
@@ -630,4 +664,285 @@ object PropertyChecker {
+      (value, Array[String](), csvwPropertyType)
+    }
+  }
+
+  def resourceProperty(csvwPropertyType: PropertyType.Value):(JsonNode, String, String) => (JsonNode, Array[String], PropertyType.Value) = {
+    (value, baseUrl, lang) => {
+      (value, Array[String](), csvwPropertyType)
+    }
+  }
+
+  def schemaReferenceProperty(csvwPropertyType: PropertyType.Value):(JsonNode, String, String) => (JsonNode, Array[String], PropertyType.Value) = {
+    (value, baseUrl, lang) => {
+      val url = baseUrl + value.asText()
+      // Don't know how to place a URI object in JsonNode, keeping the text value as of now
+      (new TextNode(url), Array[String](), csvwPropertyType)
+    }
+  }
+
+  def dialectProperty(csvwPropertyType: PropertyType.Value):(JsonNode, String, String) => (JsonNode, Array[String], PropertyType.Value) = {
+    (value, baseUrl, lang) => {
+      value match {
+        case o: ObjectNode => {
+          val valueCopy = o.deepCopy()
+          var warnings = Array[String]()
+          val fieldsAndValues = Array.from(valueCopy.fields.asScala)
+          for (fieldsAndValue <- fieldsAndValues) {
+            val key = fieldsAndValue.getKey
+            val v = fieldsAndValue.getValue
+            key match {
+              case "@id" => {
+                val matcher = PropertyChecker.startsWithUnderscore.pattern.matcher(v.asText())
+                if (matcher.matches) {
+                  throw new MetadataError("@id starts with _:")
+                }
+              }
+              case "@type" => {
+                if (v.asText() != "Dialect") {
+                  throw new MetadataError("@type of dialect is not 'Dialect'")
+                }
+              }
+              case _ => {
+                val (new_v, w, t) = checkProperty(key, v, baseUrl, lang)
+                if (t == PropertyType.Dialect && w.isEmpty) {
+                  valueCopy.set(key, new_v)
+                } else {
+                  valueCopy.remove(key)
+                  if (t != PropertyType.Dialect) {
+                    warnings = warnings :+ "invalid_property"
+                  }
+                  warnings = Array.concat(warnings, w)
+                }
+              }
+            }
+          }
+          (valueCopy, warnings, csvwPropertyType)
+        }
+        case _ => (NullNode.instance, Array[String](PropertyChecker.invalidValueWarning), csvwPropertyType)
+      }
+    }
+  }
+
+  def transformationsProperty(csvwPropertyType: PropertyType.Value):(JsonNode, String, String) => (JsonNode, Array[String], PropertyType.Value) = {
+    (value, baseUrl, lang) => {
+      var warnings = Array[String]()
+      val transformationsToReturn = JsonNodeFactory.instance.arrayNode()
+      value match {
+        case a:ArrayNode => {
+          val transformationsArr = Array.from(a.elements().asScala)
+          for((t,i) <- transformationsArr.zipWithIndex) {
+            t match {
+              case o:ObjectNode => {
```

Comment:
This is getting a bit deep and harder to understand with all the one-letter variable names. A new function may be in order here.

Reply:
Done